### PR TITLE
Add '/api/docs' endpoint to Auth server

### PR DESCRIPTION
**Pull request by kpfleming**
### Short description
Adds an '/api/docs' endpoint to allow clients to obtain the OpenAPI (Swagger) specification document directly from the server.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [x] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)
- [ ]  checked that this code was merged to master

### CODE REVIEW

**Comment by kpfleming** [comment]
There are some open issues here:

* Some clients will prefer to receive the OpenAPI document in JSON instead of YAML. There is no mechanism available in the Auth server code today to convert the YAML document into JSON, but we could configure Swagger to produce JSON instead of YAML and then supply the resulting document to clients who ask for either JSON or YAML (since JSON documents are also YAML documents).

* The new endpoint is part of the `/api` route and thus requires the api-key for access; it may be better to make the endpoint available without authentication.

* Requiring the administrator to set a configuration setting to point to the document is a bit of additional work for them, and will also require that the document be packaged and installed with the server. It may be possible to embed the document into the server binary itself so that neither of these are necessary, but I didn't see any existing examples in the code of that being done. I'm happy to make that change with some guidance for the preferred method.

**Comment by Habbie** [comment]
Thanks!


>     * we could configure Swagger to produce JSON instead of YAML and then supply the resulting document to clients who ask for either JSON or YAML (since JSON documents are also YAML documents)

That sounds like something we could do during `make dist` or package build.



>     * require that the document be packaged and installed with the server

We can ship the file in our packages.



>     * It may be possible to embed the document into the server binary itself so that neither of these are necessary, but I didn't see any existing examples in the code of that being done

dnsdist-web.cc uses `incbin` to include some static HTML, but I prefer the packaging route.

**Comment by kpfleming** [comment]
I'll also need to update the index.rst in the http-api docs to indicate that this new endpoint is available.

File: pdns/ws-api.cc
Comment (Habbie) [inline pdns/ws-api.cc]:
```suggestion
   std::ifstream ifs(fname, std::ifstream::binary);
```

pdns already requires C++11

**Comment by omoerbeek** [comment]
Please rebase, there's a change in the test setup that CircleCI picks up only partially. That explains the test-recursor-bulk test falure. Rebasing should fix that.

**Comment by kpfleming** [comment]
As far as I can tell the existing OAS (Open API Specification) document wasn't generated from any sources in this repository; is that correct? Is it being maintained by hand?

**Comment by Habbie** [comment]
> As far as I can tell the existing OAS (Open API Specification) document wasn't generated from any sources in this repository; is that correct? Is it being maintained by hand?

Yes, it is maintained by hand.

**Comment by kpfleming** [comment]
OK, so would you be willing to maintain it in JSON format instead of YAML? That would only require a one-time conversion, and no additional tools would be necessary during the packaging process.

**Comment by Habbie** [comment]
> OK, so would you be willing to maintain it in JSON format instead of YAML? That would only require a one-time conversion, and no additional tools would be necessary during the packaging process.

@pieterlexis ?

**Comment by pieterlexis** [comment]
I dislike typing JSON a lot more than YAML

> > * we could configure Swagger to produce JSON instead of YAML and then supply the resulting document to clients who ask for either JSON or YAML (since JSON documents are also YAML documents)

> That sounds like something we could do during make dist or package build.

> >    * require that the document be packaged and installed with the server

> We can ship the file in our packages.

> >   * It may be possible to embed the document into the server binary itself so that neither of these are necessary, but I didn't see any existing examples in the code of that being done

> dnsdist-web.cc uses incbin to include some static HTML, but I prefer the packaging route.

I really prefer us creating the JSON at build-time and including it in the binary. If we don't, the schema served could be changed by an operator on-disk, leading to support questions.

**Comment by kpfleming** [comment]
Indeed, that is why I proposed including it into the binary, to avoid the risk of mismatches.

There are a number of mechanisms which could be used to do the conversion; the simplest is to use Python and the PyYAML package (easily installable on any system with Python). We could also bring a C++ YAML library into the source tree to avoid external dependencies.

**Comment by Habbie** [comment]
> Indeed, that is why I proposed including it into the binary, to avoid the risk of mismatches.

Ok, agreed!



> We could also bring a C++ YAML library into the source tree to avoid external dependencies.

We use plenty of Python in our distribution and build already, so the PyYAML route sounds good to me!

**Comment by kpfleming** [comment]
Excellent, I'll update this PR with the necessary changes. Stay tuned.

**Comment by kpfleming** [comment]
New version has been posted here now. It embeds both the YAML and JSON forms of the document into the server binary, and I've tested retrieving with no Accept header, and with headers which specify application/json and application/x-yaml.

The LGTM analysis fails because it is unable to create a Python virtualenv which is required for the YAML->JSON conversion step; none of other virtualenv processes in the tree would have been triggered by LGTM previously because they are all for building documentation. We might need some help from the LGTM team figuring out how to solve this problem.

**Comment by Habbie** [comment]
We should be able to fix it in `lgtm.yml`.

**Comment by kpfleming** [comment]
It's also using Python 2.7 even though lgtm.yml explicitly asks for Python 3. I suspect the proper fix is to switch to using 'python3 -m venv .venv' instead of the virtualenv tool throughout the tree, but that's too large a change to bundle into this PR.

**Comment by Habbie** [comment]
Hack: before calling make, make an empty or sufficiently empty `apidocfiles.h` so that the compile succeeds. We never run the binary that comes out of the LGTM compile anyway.

**Comment by kpfleming** [comment]
Good call, a one-line hack in lgtm.yml is now in place.

**Comment by kpfleming** [comment]
Success... all checks are green now :-)

**Comment by zeha** [comment]
Would be great if the Recursor would support the same endpoint :)

**Comment by Habbie** [comment]
> Would be great if the Recursor would support the same endpoint :)

Worth a request ticket I suppose, I don't want to hold this PR for every product :)

**Comment by kpfleming** [comment]
Does the recursor support the same API?

**Comment by Habbie** [comment]
> Does the recursor support the same API?

No, there are a lot of differences, and we don't have a Swagger definition for it anyway.

**Comment by kpfleming** [comment]
Then I concur, an issue for that in the future seems like a good choice :-)

**Comment by bootc** [comment]
Thanks for doing this! However I don't think this is enough by itself: the server will need to substitute some values into the spec file for the `host` and `schemes`:

https://swagger.io/docs/specification/2-0/api-host-and-base-path/

When loading the spec from disk for use with Bravado I currently substitute in the values myself; it would be nice to avoid having to do this:

```
def powerdns():
    url = urlparse(cfg.powerdns.url)

    http_client = RequestsClient()
    http_client.set_api_key(
        url.netloc, cfg.powerdns.key,
        param_name='X-API-Key', param_in='header'
    )

    spec_dict = load_file('authoritative-api-swagger.yaml')
    spec_dict['host'] = url.netloc
    spec_dict['schemes'] = [url.scheme]

    return SwaggerClient.from_spec(spec_dict, http_client=http_client)
```

**Comment by kpfleming** [comment]
I do that substitution as well, for the same reason. From my understanding of the Bravado docs if `from_url` is invoked Bravado ignores the `host`, `scheme`, and `port` attributes in the spec and substitutes the ones which were used to obtain the spec. I don't know if this behavior is specific to Bravado or is common among Swagger clients.

In any case removing those attributes from the spec file would be a simple fix.

**Comment by zeha** [comment]
This PR needs a rebase to fix conflicts in ws-recursor.cc.

Should it also change the swagger spec?

**Comment by jsoref** [comment]
To help the [check-spelling](https://github.com/marketplace/actions/check-spelling) action, please rebase to `master` and add `apidocfiles` to `.github/actions/spell-check/expect.txt` as suggested:
https://github.com/jsoref/pdns/commit/e9f8a6344832bacb992c1ed5186d9cd6b95ce925#commitcomment-39229522

File: pdns/ws-recursor.cc
Comment (Habbie) [inline pdns/ws-recursor.cc:385]: why the change from `std` to `boost`?

Comment (kpfleming) [reply pdns/ws-recursor.cc:385]: I honestly have no idea where this change came from! I don't remember changing it nor can I see any reason why it would have been changed as it's not related to the topic of this PR at all.

Comment (Habbie) [reply pdns/ws-recursor.cc:385]: Probably happened during a rebase then. Can you take it out? :)

Comment (kpfleming) [reply pdns/ws-recursor.cc]: Done.

**Comment by Habbie** [comment]
This breaks when building from a distribution tarball: `builder/build.sh -m authoritative debian-buster` ends with `make[3]: *** No rule to make target '../docs/http-api/swagger/authoritative-api-swagger.yaml', needed by 'api-swagger.yaml'.  Stop.`

I can probably look at this at some later time if you don't have time or don't feel like fighting our Makefiles.

**Comment by kpfleming** [comment]
That would indicate that the file is not being included in the tarball... that's unfortunate. I hadn't looked into those portions of the Makefiles yet, so if you're able to investigate this one that would probably get to a solution more quickly.

**Comment by kpfleming** [comment]
Is this still a candidate for inclusion in 4.4.0 (since alpha releases are already being produced)? If so, I'll get it rebased and fix up the distribution tarball issue, since I've spent some time working in the builder tooling recently :-)

**Comment by Habbie** [comment]
> Is this still a candidate for inclusion in 4.4.0

Yes, within a few weeks (at beta) this is fine! Let me know if you need any help.

**Comment by kpfleming** [comment]
I'm actively working on this again, should have a new version pushed in the next 24 hours.

**Comment by kpfleming** [comment]
Branch updated; it has been rebased on current `master`, and builds from source tarballs (produced by `make dist`) now complete as they should.

**Comment by pieterlexis** [comment]
The github action config might need an update:

```
You need virtualenv to generate the JSON API document
exit 1
Makefile:3590: recipe for target 'api-swagger.json' failed
```

**Comment by kpfleming** [comment]
Ugh. It's getting harder and harder to use Python 2.7 and the `virtualenv` tool... as noted in another PR, we're long past the point where upgrading the build system to use Python 3.x is necessary.

New commit pushed which adds `python-virtualenv` package to the CodeQL workflow. Fingers crossed :-)

**Comment by kpfleming** [comment]
Another thought: since `virtualenv` is no longer optional for building the Authoritative Server, the configure script for it should bail out if it was not found, instead of proceeding and then producing an error during the build.

**Comment by Habbie** [comment]
> Another thought: since `virtualenv` is no longer optional for building the Authoritative Server, the configure script for it should bail out if it was not found, instead of proceeding and then producing an error during the build.

Is this also true when building from a tarball?

**Comment by kpfleming** [comment]
At the moment it is, because the source tarball only contains the YAML version of the Swagger document. It may be possible to have the 'dist' targets do the YAML->JSON conversion and then include both formats in the tarball, I can experiment with that.

**Comment by Habbie** [comment]
Yes, that would make sense, then it's similar to how we handle manpages (those also need Python to generate them, but then we stick them in the tarball)

**Comment by kpfleming** [comment]
OK, I'll make those changes too.

**Comment by pieterlexis** [comment]
If you need help with the autotools/autoconf/m4, let me known

**Comment by kpfleming** [comment]
OK, that's fixed. `api-swagger.yaml` and `api-swagger.json` are now included in source tarballs, so `virtualenv` is not required to build from them. If `virtualenv` is available it will still be used and changes can be made to those files as normal for inclusion into the server binary.

`virtualenv` is still required to build from a plain repository clone, though.

**Comment by zeha** [comment]
lgtm!

**Comment by kpfleming** [comment]
I dare say this might finally be ready for merging :-)

**Comment by Habbie** [comment]
Looks good at a quick skim! I've kicked off a test build at https://builder.powerdns.com/#/buildrequests/129582?redirect_to_build=true - normally we just merge and see, but as this touches the build system directly, I'd like to wait for that.

**Comment by Habbie** [comment]
@pieterlexis can you check the autotools bits in here?

**Review by pieterlexis — APPROVED** [review]
I like it

**Comment by Habbie** [comment]
@kpfleming many thanks!

**Comment by kpfleming** [comment]
You're all very welcome... now I need to update my Ansible modules which use the API to pull the Swagger spec directly from the server :-)

**Comment by kpfleming** [comment]
With the release of 4.4.0-beta1, my Ansible modules which use the API to manage Zones and TSIG Keys are now passing automated tests using direct extraction of the Swagger spec from the server, instead from a local file. Success!

https://github.com/kpfleming/ansible-pdns-auth-api

**Pull request merged**
